internal/stats/aggregate.go

```diff
@@ -0,0 +1,212 @@
+			jsonTags[jsonTag] = true
+		}
+	}
+}
+
+// Ensure the groupByFields and aggregateField exist in the metricType.
+func validate(metricType any, groupByFields []string, aggregateField string) error {
+	reflectType := reflect.ValueOf(metricType).Type()
+
+	jsonTags := make(map[string]bool)
+	extractJSONTags(reflectType, jsonTags)
+
+	for _, field := range groupByFields {
+		if !jsonTags[field] {
+			return fmt.Errorf("Failed to validate aggregation query. '%s' is not a valid group-by field.", field)
+		}
+	}
+
+	if !jsonTags[aggregateField] {
+		return fmt.Errorf("Failed to validate aggregation query. '%s' is not a valid aggregate field.", aggregateField)
+	}
+
+	return nil
+}
+
+// Group together metrics based on their groupByFields.
+func groupTogetherMetrics(metrics []map[string]any, groupByFields []string, aggregateField string) (map[string][]map[string]any, error) {
+	groupedMetricBuckets := make(map[string][]map[string]any)
+
+	for _, metric := range metrics {
+		// Skip grouping this metric if it doesn't contain all groupByFields and the aggregateField.
+		if !hasAllNeededFields(metric, groupByFields, aggregateField) {
+			continue
+		}
+
+		var groupFieldValues []string
+		for _, field := range groupByFields {
+			groupFieldValues = append(groupFieldValues, fmt.Sprintf("%v", metric[field]))
+		}
+
+		bucketKey := strings.Join(groupFieldValues, "|")
+		groupedMetricBuckets[bucketKey] = append(groupedMetricBuckets[bucketKey], metric)
+	}
+
+	return groupedMetricBuckets, nil
+}
+
+func hasAllNeededFields(metric map[string]any, groupByFields []string, aggregateField string) bool {
+	for _, field := range groupByFields {
+		_, exists := metric[field]
+		if !exists {
+			return false
+		}
+	}
+
+	_, exists := metric[aggregateField]
+	if !exists {
+		return false
+	}
+
+	return true
+}
+
+// Calculate aggregations for a group of metrics.
+func computeGroupAggregation(groupedMetrics []map[string]any, aggregateField string, groupByFields []string) (map[string]any, error) {
+	if len(groupedMetrics) == 0 {
+		return nil, nil
+	}
+
+	aggregatedResults := make(map[string]any)
+
+	// Use the first metric in the group to extract the group-by field values.
+	firstMetric := groupedMetrics[0]
+	for _, field := range groupByFields {
+		value, exists := firstMetric[field]
+		if !exists {
+			return nil, fmt.Errorf("Failed to get the group-by field '%s' for metric '%v'.", field, firstMetric)
+		}
+
+		aggregatedResults[field] = value
+	}
+
+	var values []float64
+	var totalCount int = 0
+
+	for _, metric := range groupedMetrics {
+		aggregateValue, exists := metric[aggregateField]
+		if !exists {
+			// Skip aggregating this metric if it doesn't have the aggregation field.
+			continue
+		}
+
+		totalCount++
+
+		aggregateValueString := fmt.Sprintf("%v", aggregateValue)
+		numericValue, err := util.StrToFloat(aggregateValueString)
+		// Only append to values if the aggregateValue was a number.
+		if err == nil {
+			values = append(values, numericValue)
+		}
```

Comment:
I don't like this conversion to a string and back.
Instead, I would just track for each bucket if it has a string value (thereby forcing all values to be strings).